### PR TITLE
issue #637 - changes for URL aware language tabs

### PR DESCRIPTION
**Pull request by 10xtechie**
### 🤔 What's changed?

- Switch to the appropriate language tab on opening a page if a query string parameter is provided(e.g. ?lang=java)
- Update the URL if the language tab is changed 

### ⚡️ What's your motivation? 

Implementation of an essential feature

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)

### CODE REVIEW

File: themes/cucumber-sb/static/js/site.js [hunk 1]
```diff
@@ -91,12 +91,56 @@ function showOnly(language) {
...
+  "scala"
+]
+
 ready(function() {
+  var selectedLang = getLangFromUrl();
+  if(selectedLang != '' && selectedLang != null && supportedLanguages.includes(selectedLang)){
+    showOnly(selectedLang)
+  }else{
+    //setting default language
+    showOnly("java")
```
Comment (mattwynne) [inline themes/cucumber-sb/static/js/site.js]: Might be a good idea to call this default out in a const somewhere.

**Comment by mattwynne** [comment]
This is great, thanks so much!

We could consider storing your preferred language in a cookie, and defaulting to that where the URL doesn't contain a language. I think that might help people navigating around through different pages to stay on the same language. Otherwise, I think a Ruby developer would end up seeing java examples as soon as they switch to a different page.

What do you think? Should we try and tackle that as part of this change, or add it on top in a later iteration?

**Review by mattwynne — APPROVED** [review]

**Comment by 10xtechie** [comment]
> This is great, thanks so much!
> 
> We could consider storing your preferred language in a cookie, and defaulting to that where the URL doesn't contain a language. I think that might help people navigating around through different pages to stay on the same language. Otherwise, I think a Ruby developer would end up seeing java examples as soon as they switch to a different page.
> 
> What do you think? Should we try and tackle that as part of this change, or add it on top in a later iteration?

@mattwynne ,
Good point, totally agreed. Will update accordingly :)

**Comment by 10xtechie** [comment]
@mattwynne,
All changes you suggested have been done! Please take a look :)

File: themes/cucumber-sb/static/js/site.js [hunk 2]
```diff
@@ -91,12 +91,71 @@ function showOnly(language) {
...
+
 ready(function() {
+  var selectedLang = getLangFromUrl();
+  if((selectedLang == '' || selectedLang == null)){
+    if(localStorage.getItem('language') == ''){
+      showOnly(defaulLanguage)
+      localStorage.setItem('language', defaulLanguage)
+    }else{
+      showOnly(localStorage.getItem('language'))
+    }
```
Comment (mattwynne) [inline themes/cucumber-sb/static/js/site.js:141]: Maybe these lines (136-141) which, if I'm reading it correctly, are duplicated down on line 146, could be extracted into a single function to improve readability/maintainability?

Comment (mattwynne) [reply themes/cucumber-sb/static/js/site.js:141]: I want to ship this! Could you consider adding a separate refactoring PR with this nit-pick in it?

**Comment by mattwynne** [comment]
Thanks so much for this!

One thing I notice, on testing it, is that if you move between pages, the URL loses the language, so that if you were to share the link with someone, they might see a different page to you.

I wonder if we should also rewrite the URL when we retrieve the default successfully from local storage, what do you think? Basically treat this like an automatic tab selection.

**Comment by mattwynne** [comment]
Also, I wonder if we could use https://developer.mozilla.org/en-US/docs/Web/API/History/pushState to modify the URL, to avoid a reload of the page?

**Comment by 10xtechie** [comment]
@mattwynne ,
I believe rewriting the URL while retrieving the default language successfully from local storage is a good idea, so is using pushState to modify the URL. Can I go ahead and implement them?

**Comment by aurelien-reeves** [comment]
@10xtechie You can go, yes

Please note that I had to make some changes to your code in order to fix an issue: https://github.com/cucumber/docs/pull/771

- detecting available languages is dynamic based on the tabs in the page: because all the pages have not the same available languages. [Some have scala](https://cucumber.io/docs/cucumber/step-definitions/) where [other have not](https://cucumber.io/docs/guides/10-minute-tutorial/), there is even [golang at some places](https://cucumber.io/docs/gherkin/step-organization/), to name a few.
- the default language is not hardcoded anymore
- we make sure that we always activate a language that is part of the available languages

**Pull request merged**